File: web/plugins/string_slice.js
import { WebPlugin } from "../engine/plugin.js";

export class StringSlicePlugin extends WebPlugin {
    constructor() {
        super("文字を切り出す");
    }

    async execute(line, context) {
        const match = line.match(/^文字を切り出す\s+「(.+)」\s+(.+)\s+(.+)$/);
        if (!match) return;

        const start = Number(
            context.parser.parseValue(match[2], context.runtime)
        );

        const end = Number(
            context.parser.parseValue(match[3], context.runtime)
        );

        context.runtime.write(
            match[1].slice(start, end)
        );
    }
}

File: pc/plugins/string_slice.py
import re
from core.plugin import Plugin


class StringSlicePlugin(Plugin):
    def translate(self, line):
        match = re.fullmatch(r'文字を切り出す\s+「(.+)」\s+(.+)\s+(.+)', line)
        if not match:
            return None

        text = match.group(1)
        start = match.group(2)
        end = match.group(3)

        return f"print({text!r}[{start}:{end}])"